Topics/Tech_Stacks/NoSQL_databases_JSON_interactions.md

```diff
@@ -0,0 +1,208 @@
+This parameter is not passed directly into the `find()` method like the `<query>` and `<projection>` parameters. Instead, `<options>` are specified through method chaining, where you append methods to `find()` that correspond to the various options you want to apply, such as sorting, limiting, and skipping documents.
+
+For example, you can use `sort()` to sort the results based on one or more fields.
+```
+db.collection.find().sort({ "age": 1 })
+```
+Here, we sort documents by age in ascending order.
+
+To restrict the number of documents returned, use limit().
+```
+db.collection.find().limit(5)
+```
+We limit the query to return only 5 documents here.
+
+To get a more specific result, you can combine the option method together, for example combine `limit()` and `skip()` for pagination.
+```
+db.collection.find().skip(5).limit(5)
+```
+We will get the second set of 5 documents in this case. 
+
+There are many more available options, visit here for more details.
+
+Note every parameter above is optional, to return all documents in a collection, omit all parameters or pass an empty document ({}).
+
+### Indexing JSON Data
+Indexing JSON data in a NoSQL database like MongoDB is crucial for optimizing query performance. An index in a MongoDB database is a special data structure that stores a small portion of the collection's data in an easy-to-traverse form. The index stores the value of a specific field or set of fields, ordered by the value of the field as specified in the index.
+
+The purpose of indexing is that it supports the efficient execution of queries. Without indexes, MongoDB must perform a collection scan, i.e., scan every document in a collection, to select those documents that match the query statement.
+
+There are many type of indexes in NoSQL databases:
++ Single Field: Apart from the _id field, which is automatically indexed, you can create indexes on any field in a document.
++ Compound Index: Indexes multiple fields in a single index.
++ Multikey Index: Automatically created for fields that hold arrays; used to index array elements.
++ Text Index: Created to enable text search on string content.
++ Hashed Index: Indexes the hash of the value of a field.
+Indexes have the property to be unique, which ensures that two documents cannot have the same value for the indexed field. They can be created with a specified order (ascending or descending).
+We will give some examples for some common types of indexes using createIndex method:
+
+Single Field Index: To create an index on a single field
+```
+db.collection.createIndex({ "fieldname": 1 }) // 1 for ascending order
+```
+Compound Index: To index multiple fields, specify each field and its sort order
+```
+db.collection.createIndex({ "field1": 1, "field2": -1 }) // 1 for ascending, -1 for descending
+```
+Text Index: To enable text search
+```
+db.collection.createIndex({ "fieldname": "text" })
```

Comment:
Hi chris, could you add some downside of indexing under this section? This could be helpful when people decided whether it is proper or useful to add the index when they are implementing it. Thanks!